### PR TITLE
Support processing of permission file in new format

### LINKED ISSUE #1261: Convert the current permissions model being used by DevX API to the new permissions model from Kibali.
the current permissions model here(https://github.com/microsoftgraph/microsoft-graph-devx-content/tree/dev/permissions) should be converted to the new model that looks like this (https://github.com/microsoftgraph/kibali/blob/main/test/kibaliTests/GraphPermissions.json).

This will allow us to see how DevX API will handle new permissions when Identity team start using the new model. It will also help us start tagging hidden permissions as so and thus filter them out to avoid displaying them on GE

### PR DESCRIPTION
**Pull request by millicentachieng**
## Overview

Fixes #1261, #1253, #1318, #1337, #1338, #1355, 
This PR
- Updates DevX API to use permissions file in new format
- Updates permissions endpoint to add support for returning permissions for all scope types for a given endpoint
- Updates DevX API to return scope types in the permissions response object
- Updates DevX API to return a payload indicating which permission has the least privilege
- Updates permissions endpoint to add support for filtering out hidden permissions
- Adds an endpoint that to return least privilege permissions given a set of URLs

### CODE REVIEW

File: GraphWebApi/wwwroot/OpenApi.yaml [hunk 1]
```diff
@@ -256,17 +262,128 @@ paths:
...
                   consentDescription: "Allows the app to read email in your mailbox. "
                   isAdmin: false
+                  isLeastPrivilege: false
                 - value: Mail.ReadWrite
+                  scopeType: Application
                   consentDisplayName: "Read and write access to your mail "
                   consentDescription: "Allows the app to read, update, create and delete email in your mailbox. Does not include permission to send mail. "
                   isAdmin: false
+                  isLeastPrivilege: false
+        "207":
```
Comment (baywet) [inline GraphWebApi/wwwroot/OpenApi.yaml:365]: 207 is a webdav status code. Why do we need both the scopes and the errors at the same time? shouldn't it be one or the other?

File: PermissionsService/Interfaces/IPermissionStore.cs [hunk 2]
```diff
@@ -17,19 +17,40 @@ public interface IPermissionsStore
...
+        /// </summary>
+        /// <param name="requestUrls">The list of request URLs to fetch permissions for.</param>
+        /// <param name="scopeType">Optional: The type of scope to be retrieved for the target request url.</param>
+        /// <param name="locale">Optional: The language code for the preferred localized file.</param>
+        /// <param name="org">Optional: The name of the org/owner of the repo.</param>
+        /// <param name="branchName">Optional: The name of the branch containing the files</param>
+        /// <param name="includeHidden">Optional: Whether to include hidden permissions or not: Defaults to false.</param>
+        /// <param name="isLeastPrivilege">Optional: Whether to only return least privilege permissions on not. Defaults to true.</param>
+        /// <returns></returns>
+        Task<PermissionResult> GetScopesForMultipleResourcesAsync(List<string> requestUrls = null,
```
Comment (baywet) [inline PermissionsService/Interfaces/IPermissionStore.cs:47]: instead of introducing a new method, since we own both the interface and the implementation, why don't we evolve the existing one?

File: PermissionsService/Models/PathPermissions.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace PermissionsService.Models
+{
+    internal class PathPermissions
+    {
+        public string Url
```
Comment (baywet) [inline PermissionsService/Models/PathPermissions.cs:8]:
```suggestion
        public required string Url
```

(this way you can get rid of the constructor)

File: PermissionsService/Models/PathSet.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        [JsonPropertyName("paths")]
+        [JsonConverter(typeof(PathsLeastPrivilegeSchemeKeyConverter))]
+        public Dictionary<string, List<ScopeType>> Paths
+        {
+            get; set;
+        } = new();
+    }
+
+
+    internal class PathsLeastPrivilegeSchemeKeyConverter : JsonConverter<Dictionary<string, List<ScopeType>>>
```
Comment (baywet) [inline PermissionsService/Models/PathSet.cs:47]: why do we need custom converters, couldn't we adapt our models to avoid needing them?

File: PermissionsService/Services/PermissionsStore.cs [hunk 5]
```diff
@@ -94,79 +95,108 @@ private PermissionsDataInfo PermissionsData
...
                         }
                     }
                 }
             }
 
-            _telemetryClient?.TrackTrace("Finished seeding permissions tables",
-                                         SeverityLevel.Information,
-                                         _permissionsTraceProperties);
+            var hiddenPermissions = permissionsDocument.Permissions
+                .Where(x => x.Value.ProvisioningInfo.IsHidden).Select(x => x.Key).ToHashSet();
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:193]:
```suggestion
                .Where(static x => x.Value.ProvisioningInfo.IsHidden).Select(static x => x.Key).ToHashSet(StringComparer.OrdinalIgnoreCase);
```

File: PermissionsService/Services/PermissionsStore.cs [hunk 6]
```diff
@@ -319,111 +349,147 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
+                                                           string branchName = null,
+                                                           bool includeHidden = false,
+                                                           bool isLeastPrivilege = true)
+        {
+            if (PermissionsData?.PathPermissions == null)
+                throw new InvalidOperationException("Failed to fetch permissions");
 
-                scopesInfo = GetScopesInformation(scopesInformationDictionary, scopes, scopeType, true);
+            // Create a dict of scopes information from GitHub files or cached files
+            var scopesInformationDictionary = (!string.IsNullOrEmpty(org) && !string.IsNullOrEmpty(branchName))
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:396]:
```suggestion
            var scopesInformationDictionary = !(string.IsNullOrEmpty(org) || string.IsNullOrEmpty(branchName))
```

File: PermissionsService/Services/PermissionsStore.cs [hunk 7]
```diff
@@ -319,111 +349,147 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
+                Errors = errors
+            };
+        }
 
-                if (permissionsData.ScopesListTable[key] is not JToken resultValue)
+        private IEnumerable<ScopeInformation> GetAllScopes(ScopeType? scopeType, bool includeHidden)
+        {
+            var scopes = PermissionsData.PathPermissions.Values
+                .SelectMany(x => x.SupportedMethodPermissions.Values)
+                .SelectMany(x => x)
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:438]:
```suggestion
                .SelectMany(static x => x.SupportedMethodPermissions.Values)
                .SelectMany(static x => x)
```

File: PermissionsService/Services/PermissionsStore.cs [hunk 8]
```diff
@@ -319,111 +349,147 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
+                .Where(x => scopeType == null || x.Key == scopeType)
+                .SelectMany(x => x.Value.AllPermissions, (x, permission) => new ScopeInformation
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:440]:
```suggestion
                .SelectMany(static x => x.Value.AllPermissions, static (x, permission) => new ScopeInformation
```

**Review by baywet — CHANGES_REQUESTED** [review]

File: GraphWebApi/wwwroot/OpenApi.yaml (hunk 1 above)
Comment (millicentachieng) [reply GraphWebApi/wwwroot/OpenApi.yaml]: Noted. I've updated the status code to 400. We need to let the client know that there were some URLs that did not yield  any scopes and they can proceed with the result if they choose to.

File: PermissionsService/Interfaces/IPermissionStore.cs (hunk 2 above)
Comment (millicentachieng) [reply PermissionsService/Interfaces/IPermissionStore.cs]: Done. I've merged the two methods.

File: PermissionsService/Models/PathSet.cs (hunk 4 above)
Comment (millicentachieng) [reply PermissionsService/Models/PathSet.cs]: I've removed the custom converters.

File: GraphWebApi/wwwroot/OpenApi.yaml [hunk 9]
```diff
@@ -745,18 +859,24 @@ components:
...
+        - scopeType
         - consentDisplayName
         - consentDescription
         - isAdmin
+        - isLeastPrivilege
       properties:
         value:
           type: string
+        scopeType:
+          type: string
```
Comment (baywet) [inline GraphWebApi/wwwroot/OpenApi.yaml:871]: this could use an enum type instead

File: GraphWebApi/wwwroot/OpenApi.yaml [hunk 10]
```diff
@@ -256,23 +262,130 @@ paths:
...
+      operationId: permissions.ListPermissionsSet
+      parameters:
+        - name: scopeType
+          in: query
+          schema:
+            type: string
+            enum:
+              - DelegatedWork
+              - DelegatedPersonal
+              - Application
```
Comment (baywet) [inline GraphWebApi/wwwroot/OpenApi.yaml:302]: why not move this definition to the schema + parameters components and refer to it here instead to enable reusability?

File: GraphWebApi/wwwroot/OpenApi.yaml [hunk 11]
```diff
@@ -256,23 +262,130 @@ paths:
...
+          required: false
+          description: If set to true, allows fetching of any available hidden permissions
+          default: false
+        - name: isLeastPrivilege
+          in: query
+          schema:
+            type: boolean
+          required: false
+          description: If set to true, fetches minimum permissions for a set of resources
+          default: true
```
Comment (baywet) [inline GraphWebApi/wwwroot/OpenApi.yaml:330]: parameters components for all of them so they can be reused between the get and post operations

File: PermissionsService/Models/PathPermissions.cs [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+            get; set;
+        } = new(StringComparer.OrdinalIgnoreCase);
+    }
+
+    internal class SchemePermissions
+    {
+        public HashSet<string> LeastPrivilegePermissions
+        {
+            get; set;
+        } = new();
```
Comment (baywet) [inline PermissionsService/Models/PathPermissions.cs:24]:
```suggestion
        } = new(StringComparer.OrdinalIgnoreCase);
```

File: PermissionsService/Models/PathPermissions.cs [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+        public HashSet<string> AllPermissions
+        {
+            get; set;
+        } = new();
```
Comment (baywet) [inline PermissionsService/Models/PathPermissions.cs:29]:
```suggestion
        } = new(StringComparer.OrdinalIgnoreCase);
```

File: PermissionsService/Models/PathSet.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace PermissionsService.Models
+{
+    internal class PathSet
+    {
+        [JsonPropertyName("schemeKeys")]
+        public HashSet<string> SchemeKeys
+        {
+            get; set;
+        } = new();
```
Comment (baywet) [inline PermissionsService/Models/PathSet.cs:13]:
```suggestion
        } = new(StringComparer.OrdinalIgnoreCase);
```

File: PermissionsService/Models/PathSet.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+        [JsonPropertyName("methods")]
+        public HashSet<string> Methods
+        {
+            get; set;
+        } = new(StringComparer.OrdinalIgnoreCase);
+
+        [JsonPropertyName("excludedProperties")]
+        public List<string> ExcludedProperties
```
Comment (baywet) [inline PermissionsService/Models/PathSet.cs:22]: any specific reason why we're using a list and not a hashset for those two?

File: PermissionsService/Models/PermissionResult.cs [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+using System.Net;
+
+namespace PermissionsService.Models
+{
+    public class PermissionResult
+    {
+        public List<ScopeInformation> Results
+        {
+            get; set;
+        }
```
Comment (baywet) [inline PermissionsService/Models/PermissionResult.cs:11]:
```suggestion
        } = new();
```

File: PermissionsService/Models/PermissionResult.cs [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+
+        public List<PermissionError> Errors
+        {
+            get; set;
+        }
```
Comment (baywet) [inline PermissionsService/Models/PermissionResult.cs:16]:
```suggestion
        } = new();
```

File: PermissionsService/Models/PermissionResult.cs [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    public class PermissionError
+    {
+        public string Url
+        {
+            get; set;
+        }
```
Comment (baywet) [inline PermissionsService/Models/PermissionResult.cs:24]:
```suggestion
        } = string.Empty;
```

File: PermissionsService/Models/PermissionResult.cs [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+        public string Message
+        {
+            get; set;
+        }
```
Comment (baywet) [inline PermissionsService/Models/PermissionResult.cs:29]:
```suggestion
        } = string.Empty;
```

Should we decide to enable NRT for this project, we won't have to revisit this class this way

File: PermissionsService/Services/PermissionsStore.cs [hunk 20]
```diff
@@ -47,23 +45,27 @@ public class PermissionsStore : IPermissionsStore
...
+            public IDictionary<string, PathPermissions> PathPermissions
             {
-                get;set;
+                get; set;
             }
 
+            public HashSet<string> HiddenPermissions
+            {
+                get; set;
+            }
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:66]: set default values?

File: PermissionsService/Services/PermissionsStore.cs [hunk 21]
```diff
@@ -94,79 +93,121 @@ private PermissionsDataInfo PermissionsData
                 {
                     lock (_permissionsLock)
                     {
-                        permissionsData = LoadPermissionsData();
-                        _cache.Set("PermissionsData", permissionsData, new MemoryCacheEntryOptions
+                        permissionsData = LoadPermissionsDataAsync().Result;
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:96]: instead of using a lock and blocking everyone, use [GetOrCreateAsync](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.memory.cacheextensions.getorcreateasync?view=dotnet-plat-ext-7.0) and change the property type to be a Task<PermissionsDataInfo> so everyone can await, which will be deferred in most cases and avoid blocking everything, everywhere.

**Review by baywet — CHANGES_REQUESTED** [review]

File: PermissionsService/Models/PathSet.cs (hunk 15 above)
Comment (millicentachieng) [reply PermissionsService/Models/PathSet.cs:22]: Updating to HashSet

File: PermissionsService/Models/ScopeType.cs [hunk 22]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (irvinesunday) [inline PermissionsService/Models/ScopeType.cs:1]: Remove unnecessary usings

Comment (irvinesunday) [reply PermissionsService/Models/ScopeType.cs:1]: Add license info.

File: PermissionsService/Models/ScopeType.cs [hunk 23]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PermissionsService.Models
+{
+    public enum ScopeType
```
Comment (irvinesunday) [inline PermissionsService/Models/ScopeType.cs:9]: Include public XML summary

File: GraphWebApi/Controllers/PermissionsController.cs [hunk 24]
```diff
@@ -39,53 +42,78 @@ public PermissionsController(IPermissionsStore permissionsStore, TelemetryClient
...
+                                                             [FromQuery]string branchName = null,
+                                                             [FromQuery]bool includeHidden = false,
+                                                             [FromQuery]bool isLeastPrivilege = false)
         {
-            string localeCode = RequestHelper.GetPreferredLocaleLanguage(Request) ?? Constants.DefaultLocale;
-            _telemetryClient?.TrackTrace($"Request to fetch permissions for locale '{localeCode}'",
-                                            SeverityLevel.Information,
-                                            _permissionsTraceProperties);
+            if (!string.IsNullOrEmpty(requestUrl) && string.IsNullOrEmpty(method))
+                return BadRequest("The HTTP method value cannot be null or empty.");
```
Comment (irvinesunday) [inline GraphWebApi/Controllers/PermissionsController.cs:54]: Exception message does not indicate whether the `requestUrl` could be null or empty

File: GraphWebApi/Controllers/PermissionsController.cs [hunk 25]
```diff
@@ -39,53 +42,78 @@ public PermissionsController(IPermissionsStore permissionsStore, TelemetryClient
...
-                                                             [FromQuery]string branchName = null)
+                                                             [FromQuery]string branchName = null,
+                                                             [FromQuery]bool includeHidden = false,
+                                                             [FromQuery]bool isLeastPrivilege = false)
         {
-            string localeCode = RequestHelper.GetPreferredLocaleLanguage(Request) ?? Constants.DefaultLocale;
-            _telemetryClient?.TrackTrace($"Request to fetch permissions for locale '{localeCode}'",
-                                            SeverityLevel.Information,
-                                            _permissionsTraceProperties);
+            if (!string.IsNullOrEmpty(requestUrl) && string.IsNullOrEmpty(method))
```
Comment (irvinesunday) [inline GraphWebApi/Controllers/PermissionsController.cs:53]: What if the `requestUrl` is empty? Do we want to throw an error?

File: GraphWebApi/Controllers/PermissionsController.cs [hunk 26]
```diff
@@ -39,53 +42,78 @@ public PermissionsController(IPermissionsStore permissionsStore, TelemetryClient
...
+            return result == null || !result.Results.Any() ? NotFound() : Ok(result.Results);
+        }
 
-            if (!string.IsNullOrEmpty(org) && !string.IsNullOrEmpty(branchName))
-            {
-                // Fetch permissions descriptions file from Github
-                result = await _permissionsStore.GetScopesAsync(scopeType: scopeType,
+        [HttpPost]
+        [Produces("application/json")]
+        public async Task<IActionResult> GetPermissionScopes([FromBody] List<string> requestUrls,
```
Comment (irvinesunday) [inline GraphWebApi/Controllers/PermissionsController.cs:78]: Check against nullability of `requestUrls`.

File: GraphWebApi/Controllers/PermissionsController.cs [hunk 27]
```diff
@@ -39,53 +42,78 @@ public PermissionsController(IPermissionsStore permissionsStore, TelemetryClient
...
+                                                             [FromQuery] ScopeType? scopeType = null,
```
Comment (irvinesunday) [inline GraphWebApi/Controllers/PermissionsController.cs:79]: Same as above, check for nullability.

File: PermissionsService/Services/PermissionsStore.cs [hunk 28]
```diff
@@ -94,79 +93,121 @@ private PermissionsDataInfo PermissionsData
...
             {
-                _telemetryClient?.TrackTrace($"Seeding permissions table from file source '{permissionFilePath}'",
+                var permissionsSourceUrl = _configuration.GetValue<string>("PermissionsSourceUrl");
+                _telemetryClient?.TrackTrace($"Fetching permissions from file source '{permissionsSourceUrl}'",
                                              SeverityLevel.Information,
                                              _permissionsTracePropertiesWithSanitizeIgnore);
 
-                string relativePermissionPath = FileServiceHelper.GetLocalizedFilePathSource(_permissionsContainerName, permissionFilePath);
-                string jsonString = _fileUtility.ReadFromFile(relativePermissionPath).GetAwaiter().GetResult();
+                using var client = new HttpClient();
```
[the remark this answers is not in the document]
Comment (irvinesunday) [reply PermissionsService/Services/PermissionsStore.cs:125]: Or use the already declared http client private field.
```CSharp 
private readonly IHttpClientUtility _httpClientUtility;
```

File: GraphWebApi/Controllers/PermissionsController.cs (hunk 25 above)
Comment (millicentachieng) [reply GraphWebApi/Controllers/PermissionsController.cs:53]: If `requestUrl` is null or empty then we return all permissions. We won't need to throw an error.

File: GraphWebApi/Controllers/PermissionsController.cs [hunk 29]
```diff
@@ -39,53 +42,78 @@ public PermissionsController(IPermissionsStore permissionsStore, TelemetryClient
         // Gets the permissions scopes
         [HttpGet]
         [Produces("application/json")]
-        public async Task<IActionResult> GetPermissionScopes([FromQuery]string scopeType = "DelegatedWork",
+        public async Task<IActionResult> GetPermissionScopes([FromQuery]ScopeType? scopeType = null,
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply GraphWebApi/Controllers/PermissionsController.cs:45]: We want it to be nullable so that if `scopeType` is not specified then we return all scope types.

File: PermissionsService/Services/PermissionsStore.cs [hunk 30]
```diff
@@ -319,112 +360,133 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
-                // Creates a dict of scopes information from cached files
-                scopesInformationDictionary = await GetOrCreatePermissionsDescriptionsAsync(locale);
-            }
-
-            var scopesInfo = new List<ScopeInformation>();
+            List<ScopeInformation> scopesInfo = new List<ScopeInformation>();
+            List<PermissionError> errors = new List<PermissionError>();
 
-            if (string.IsNullOrEmpty(requestUrl))  // fetch all permissions
+            if (!requestUrls.Any())
```
Comment (irvinesunday) [inline PermissionsService/Services/PermissionsStore.cs:383]: `requestUrls` could be be null here.

File: PermissionsService/Services/PermissionsStore.cs [hunk 31]
```diff
@@ -319,112 +360,133 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
 
-                    return null;
-                }
+        private IEnumerable<ScopeInformation> GetScopesForRequestUrl(string requestUrl,
+                                                              string method = null,
+                                                              ScopeType? scopeType = null,
+                                                              bool includeHidden = false,
+                                                              bool isLeastPrivilege = false)
+        {
+            if (string.IsNullOrEmpty(requestUrl))
```
Comment (irvinesunday) [inline PermissionsService/Services/PermissionsStore.cs:452]: Consider moving this check to the calling method and check whether the `IEnumerable` of `requestUrls` is null.

File: PermissionsService/Services/PermissionsStore.cs [hunk 32]
```diff
@@ -319,112 +360,133 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
+                // Get all scopes
+                var allScopes = GetAllScopes(scopeType, includeHidden);
+                scopesInfo.AddRange(allScopes);
             }
-            else // fetch permissions for a given request url and method
+            else
             {
-                if (string.IsNullOrEmpty(method))
+                // Get scopes for multiple request URLs
+                foreach (var url in requestUrls)
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs:392]: The thrown exception is handled within the catch block and execution continues to the next URL. And what are the benefits of using a select statement instead?

File: PermissionsService/Services/PermissionsStore.cs [hunk 33]
```diff
@@ -94,79 +93,121 @@ private PermissionsDataInfo PermissionsData
...
+            catch (Exception exception)
+            {
+                _telemetryClient?.TrackException(exception);
+                permissionsDocument = null;
+            }
+            return permissionsDocument;
+        }
 
-                    JToken apiPermissions = permissionsObject.First.First;
+        private async Task<PermissionsDataInfo> LoadPermissionsDataAsync()
```
Comment (zengin) [inline PermissionsService/Services/PermissionsStore.cs:141]: I think we should run this conversion once at the time of permission file deployment and generate the reverse lookup data structure on file.

DevX API should simply load that reverse lookup file as a data structure and run lookup logic per request.

Current architecture has a very long latency for the initial request which can be avoided.

**Review by zengin — CHANGES_REQUESTED** [review]

File: GraphWebApi/Controllers/PermissionsController.cs (hunk 27 above)
Comment (irvinesunday) [reply GraphWebApi/Controllers/PermissionsController.cs:79]: Resolving this as it has been explained above.

File: PermissionsService/Services/PermissionsStore.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (baywet) [reply PermissionsService/Services/PermissionsStore.cs:125]: @irvinesunday is there a specific reason why we rolled out our "own version" of the http client factory?

File: PermissionsService/Services/PermissionsStore.cs (hunk 32 above)
[the remark this answers is not in the document]
Comment (baywet) [reply PermissionsService/Services/PermissionsStore.cs:392]: I might be missing some of the UX context here. My assumption was that as an application developer, I want to see all the permissions I might need for my application to run. If there's is a failure, instead of getting only the ones that didn't fail until the first failure, wouldn't it be better to get all the ones that didn't fail at least?

File: PermissionsService/Services/PermissionsStore.cs [hunk 34]
```diff
@@ -94,79 +93,121 @@ private PermissionsDataInfo PermissionsData
...
+        {
+            var permissionsDocument = await LoadPermissionsDocumentAsync();
+            if (permissionsDocument == null)
+                return null;
 
-                    foreach (JProperty property in apiPermissions.OfType<JProperty>())
+            var uriTemplateMatcher = new UriTemplateMatcher();
+            var pathPermissions = new Dictionary<string, PathPermissions>();
+            foreach (var permission in permissionsDocument.Permissions)
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs:149]: For performance improvement, we will implement this [idea](https://github.com/microsoftgraph/microsoft-graph-devx-api/pull/1402#discussion_r1128667816) from @zengin.

File: PermissionsService/Services/PermissionsStore.cs (hunk 33 above)
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs:141]: Here's the link to the reverse lookup file I'll be using 
https://raw.githubusercontent.com/microsoftgraph/microsoft-graph-devx-content/task/ma/add-reverse-lookup-permission-file/permissions/permissions-all.json

File: PermissionsService/Services/PermissionsStore.cs (hunk 32 above)
[the remark this answers is not in the document]
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs:392]: Here's a snippet of more of the code
```csharp
foreach (var url in requestUrls)
{
    try
    {
        if (string.IsNullOrEmpty(url))
            throw new InvalidOperationException("The request URL cannot be null or empty.");

        var scopesForUrl = GetScopesForRequestUrl(requestUrl: url,
                                                scopeType: scopeType,
                                                includeHidden: includeHidden,
                                                leastPrivilegeOnly: leastPrivilegeOnly);

        if (scopesForUrl == null)
            throw new InvalidOperationException($"Permissions information for {url} were not found.");

        scopesInfo.AddRange(scopesForUrl);
    }
    catch (Exception exception)
    {
        errors.Add(new PermissionError()
        {
            Url = url,
            Message = exception.Message
        });
    }
}
```
If there's an exception while fetching scopes for any of the Urls, it's caught, added to the errors list and we move on to process the next Url.

File: PermissionsService/Services/PermissionsStore.cs (hunk 33 above)
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs]: Updated to use reverse lookup file

File: PermissionsService/Services/PermissionsStore.cs (hunk 32 above)
[the remark this answers is not in the document]
Comment (baywet) [reply PermissionsService/Services/PermissionsStore.cs:344]: thanks for clarifying/updating.

File: PermissionsService/Services/PermissionsStore.cs [hunk 35]
```diff
@@ -319,111 +312,146 @@ private async Task<string> FetchHttpSourceDocument(string sourceUri)
...
+                    }
                 }
+            }
 
-                requestUrl = CleanRequestUrl(requestUrl);
-
-                // Check if requestUrl is contained in our Url Template table
-                TemplateMatch resultMatch = permissionsData.UriTemplateMatcher.Match(new Uri(requestUrl, UriKind.RelativeOrAbsolute));
+            // Get consent display name and description
+            var scopesInfo = GetAdditionalScopesInformation(scopesInformationDictionary, scopes.DistinctBy(x => new { x.ScopeName, x.ScopeType }));
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:373]:
```suggestion
            var scopesInfo = GetAdditionalScopesInformation(scopesInformationDictionary, scopes.DistinctBy(x => $"{x.ScopeName}{x.ScopeType}", StringComparison.OrdinalIgnoreCase));
```

Otherwise you're doing reference comparisons on anonymous objects which will always be different

File: GraphWebApi/Controllers/PermissionsController.cs [hunk 36]
```diff
@@ -39,53 +42,81 @@ public PermissionsController(IPermissionsStore permissionsStore, TelemetryClient
...
+
+            _permissionsTraceProperties.Add(UtilityConstants.TelemetryPropertyKey_SanitizeIgnore, nameof(PermissionsController));
+            _telemetryClient?.TrackTrace($"Fetched {result?.Results.Count ?? 0} permissions",
                                             SeverityLevel.Information,
                                             _permissionsTraceProperties);
-                localeCode = supportedLocaleCode;
-            }
 
-            List<ScopeInformation> result = null;
+            return result == null || !result.Results.Any() ? NotFound() : Ok(result.Results);
```
Comment (baywet) [inline GraphWebApi/Controllers/PermissionsController.cs:73]: why is there a difference in returning status based on the result and presence of errors or not between the two methods?

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for making the changes, two minor things on my side

File: PermissionsService/Services/PermissionsStore.cs (hunk 35 above)
Comment (baywet) [reply PermissionsService/Services/PermissionsStore.cs:373]: right, this ends up calling [Object.Equals()](https://learn.microsoft.com/en-us/dotnet/api/system.object.equals?view=net-7.0) which can have a bunch of size effects and relies on reflection (bad perf) which recursively does reference comparisons on the object structure.
Whereas a string is deterministic and will provide better perf

File: GraphWebApi/Controllers/PermissionsController.cs (hunk 36 above)
Comment (millicentachieng) [reply GraphWebApi/Controllers/PermissionsController.cs:73]: Implemented the above

**Review by baywet — DISMISSED** [review]
Thanks for making the changes

File: PermissionsService.Test/PermissionsService.Test.csproj [hunk 37]
```diff
@@ -35,15 +34,12 @@
...
-      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
-    </Content>
-    <Content Include="TestFiles\permissions-test-file-v1.0_ver1.json">
-      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
-    </Content>
-    <Content Include="TestFiles\permissions-test-file-v1.0_ver2.json">
+    <Content Include="TestFiles\permissions-test-file-empty.json">
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </Content>
+      <Content Include="TestFiles\permissions-test-file.json">
```
Comment (zengin) [inline PermissionsService.Test/PermissionsService.Test.csproj:40]: nit: you can use wildcards to copy files, e.g. `TestFiles\*`

File: PermissionsService/Interfaces/IPermissionStore.cs [hunk 38]
```diff
@@ -15,21 +15,25 @@ namespace PermissionsService.Interfaces
...
-        /// <returns>A list of <see cref="ScopeInformation"/>.</returns>
-        Task<List<ScopeInformation>> GetScopesAsync(string scopeType = "DelegatedWork",
-                                                    string locale = null,
-                                                    string requestUrl = null,
-                                                    string method = null,
-                                                    string org = null,
-                                                    string branchName = null);
+        /// <param name="branchName">Optional: The name of the branch containing the files</param>
+        /// <returns></returns>
+        Task<PermissionResult> GetScopesAsync(List<string> requestUrls = null,
```
Comment (zengin) [inline PermissionsService/Interfaces/IPermissionStore.cs:29]: I think requestUrls should not be optional.

**Review by zengin — APPROVED** [review]

File: PermissionsService/Interfaces/IPermissionStore.cs (hunk 38 above)
Comment (millicentachieng) [reply PermissionsService/Interfaces/IPermissionStore.cs:29]: We want to fetch all permissions when requestUrls is null

File: PermissionsService.Test/PermissionsService.Test.csproj (hunk 37 above)
Comment (millicentachieng) [reply PermissionsService.Test/PermissionsService.Test.csproj]: Updated

**Review by baywet — APPROVED** [review]

**Review by baywet — CHANGES_REQUESTED** [review]

File: PermissionsService/Services/PermissionsStore.cs [hunk 39]
```diff
@@ -479,33 +504,32 @@ private static string CleanRequestUrl(string requestUrl)
                 throw new ArgumentNullException(nameof(scopes));
             }
 
-            if (string.IsNullOrEmpty(scopeType))
-            {
-                throw new ArgumentException($"'{nameof(scopeType)}' cannot be null or empty.", nameof(scopeType));
-            }
+            var schemeKeys = scopes.Select(x => x.ScopeType).Distinct()
+                .Select(x => x.ToString().Contains(Delegated, StringComparison.InvariantCultureIgnoreCase) ? Delegated : Application)
```
[the remark this answers is not in the document]
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs:508]: The descriptions list is grouped into two lists: "Delegated" and "Application". We are comparing an enum and a string that do not exactly match.

[the remark this answers is not in the document]
Comment (baywet) [reply PermissionsService/Services/PermissionsStore.cs:508]: So you're mapping two values to two other values? A dictionary would be much faster at doing that.

[the remark this answers is not in the document]
Comment (baywet) [reply PermissionsService/Services/PermissionsStore.cs:508]: or more accurately 3 enum values to 2 strings. so a Dictionary<ScopeType, string> with the 3 entries as a static readonly field shuold do the trick

File: PermissionsService/Services/PermissionsStore.cs [hunk 40]
```diff
@@ -78,97 +79,83 @@ public UriTemplateMatcher UriTemplateMatcher
...
             _scopesInformation = configuration[ScopesInfoBlobConfig]
                 ?? throw new ArgumentNullException(nameof(ScopesInfoBlobConfig), $"Config path missing: {ScopesInfoBlobConfig}");
             _defaultRefreshTimeInHours = FileServiceHelper.GetFileCacheRefreshTime(configuration[CacheRefreshTimeConfig]);
-
         }
 
-        private PermissionsDataInfo PermissionsData
-        {
-            get
+        private Task<PermissionsDataInfo> PermissionsData =>
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:89]: could you log an issue about multiple threads potentially populating the value concurrently and the impact on the initial request. And the suggested solution to load the cache on app startup instead + have something else re-populate it on a schedule.

**Review by baywet — APPROVED** [review]

File: PermissionsService/Services/PermissionsStore.cs (hunk 40 above)
Comment (millicentachieng) [reply PermissionsService/Services/PermissionsStore.cs:89]: https://github.com/microsoftgraph/microsoft-graph-devx-api/issues/1456

File: PermissionsService/Services/PermissionsStore.cs [hunk 41]
```diff
@@ -479,40 +507,45 @@ private static string CleanRequestUrl(string requestUrl)
...
+                }
+            });
 
             var scopesInfo = scopes.Select(scope =>
             {
-                scopesInformationDictionary[key].TryGetValue(scope, out var scopeInfo);                
-                return scopeInfo ?? new() { ScopeName = scope };
+                permissionDescriptionGroups.TryGetValue((ScopeType)scope.ScopeType, out var schemeKey);
+                scopesInformationDictionary[schemeKey].TryGetValue(scope.ScopeName, out var scopeInfo);
+                if (scopeInfo != null)
```
Comment (baywet) [inline PermissionsService/Services/PermissionsStore.cs:526]:
```suggestion
                if (scopesInformationDictionary[schemeKey].TryGetValue(scope.ScopeName, out var scopeInfo))
```

**Review by baywet — CHANGES_REQUESTED** [review]
